### ISSUE TITLE
connection rate increase significantly after upgrade to 1.77.0

### ISSUE THREAD

**Issue by eddietanshuo** [VictoriaMetrics/VictoriaMetrics#2552]
**Describe the bug**
Hello @valyala, we find a new problem with the conn pool.  To solve the issue https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2508, we upgrade the vmselect to newest build. It seems the tcp conn rate increased from 10 - 1k.

The cluster **without** upgrading we have the monitor:
![image](https://user-images.githubusercontent.com/103412360/167234341-5c74d875-8522-4b1b-907a-4d1bb02fb6c0.png)

The cluster **with** upgrading we have the monitor:
![image](https://user-images.githubusercontent.com/103412360/167234384-c353f658-dac7-4eaa-951e-0bba4d6445e4.png)

We also observed spikes with tcp connections after upgrade:
![image](https://user-images.githubusercontent.com/103412360/167234431-3bbe72c3-9494-4d52-baa2-a61b132c875d.png)

The query request rate is very stable:
![image](https://user-images.githubusercontent.com/103412360/167234480-62e48f96-2c7a-4cc8-b4e6-6051757bc75c.png)

**Suggestion**
It's possible that the `put()` function will put more connections than the real active connection. For example, we have an empty connection pool at begin,  when vmselect receive request, it will call "get()" function immediately, and then it will dial more than 2k connections since the first batch of request not return the connection. After this period, we left too many  starved connections which will cleanup by the `closeIdleConns` function.    

It's also possible that the query requests are not uniform-distributed. `closeIdleConns` will close most connections so that not ready to handle next spike. 

I suggest to remain a `minIdleConns` inside a connection pool. After vmselect run at init stage, it could dial n connections and put them into the connection pool. We start a runloop to check if the connections more than n, if more than n, we then close timeout connections to let connections reach n.

**Comment by eddietanshuo** [comment]
For the metrics of idle connections inside the pool, we see conn_pool reserves too many connections.
![image](https://user-images.githubusercontent.com/103412360/167412150-786a6630-7f62-4670-980d-80886e6a7e0b.png)


For the shake of high num of connections. The goroutine num increases.
![image](https://user-images.githubusercontent.com/103412360/167412746-d167cbd6-3f71-486f-912f-a01e017d3338.png)

It will result in high gc pause and high load avg.
![image](https://user-images.githubusercontent.com/103412360/167413305-dbff09ec-255c-429b-bfcc-db69dcb975a9.png)
![image](https://user-images.githubusercontent.com/103412360/167413900-82815c37-0606-4503-82db-3feb63b9a7d1.png)

**Comment by valyala** [comment]
@eddietanshuo , could you build `vmselect` from the commit d8a276fbe471527b8da43a0fb7cb92321b2d982e according to [these docs](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#building-from-sources) and verify whether this helps reducing connection churn rate for your case?

Regarding the `-search.maxConcurrentRequests` command-line flag at `vmselect` nodes mentioned by @f41gh7  above. It is  limited by 16 on purpose: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/e35e4882206a0b0e610867708d01c49138b1775f/app/vmselect/main.go#L52-L64

It should limit the number of concurrent requests, which, in turn, should limit the number of concurrent connections from vmselect to vmstorage to quite low number. According to the screenshots above, every vmselect establishes up to 200K concurrent connections to vmstorage nodes on your workload. If you have 8 vmstorage nodes according to https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2508 , then each vmselect establishes up to 200K/8=25K concurrent connections per each vmstorage nodes. This is unexpectedly high number given the 16 concurrent queries limit. This means that every concurrent query may establish 25K/16=1563 connections per each vmstorage node. A simple query with a single [time series selector](https://prometheus.io/docs/prometheus/latest/querying/basics/#time-series-selectors) needs only a single connection to vmstorage. There may be the following issues with your setup, which leads to high number of concurrent connections:

1) There are complex queries with more than a thousand of time series selectors, which lead to establishing thousands of connections from vmselect to vmstorage per query.
2) There is a bug in `vmselect`, which sometimes leads to thousands of concurrent connections to vmstorage under normal workload.

**Comment by eddietanshuo** [comment]
Sorry for late reply.
I read the source code of vmselect's netstorage and understand how it works. 
@valyala is right: we didn't set the search concurrent args, but we have a very slow query which contains 100 subqueries join by 'or' statement.  
We try to optimize such kind of query to put them into single time series selector. 

Thanks for help again.

**Comment by eddietanshuo** [comment]
I find there is a query with more than 10k subquery... this query must be the problem

**Comment by eddietanshuo** [comment]
I build vmselect with https://github.com/VictoriaMetrics/VictoriaMetrics/commit/d8a276fbe471527b8da43a0fb7cb92321b2d982e 

one of node still has high rate of connections:
![image](https://user-images.githubusercontent.com/103412360/168272403-0ba235ba-8f28-4eca-ae85-eaa708b8d2b7.png)
![image](https://user-images.githubusercontent.com/103412360/168272789-4ffaca8d-ecac-4d31-9554-363d5d6e4344.png)

**Comment by eddietanshuo** [comment]
the `concurrentDialsCh` only limits the rate to `handshakeFunc`, is there anyway to limit the max num of connection?

**Comment by eddietanshuo** [comment]
By the way, I find some goroutine are blocked by pacelimiter with conditional variable. They seems never wakeup:
goroutine 1010 [sync.Cond.Wait, 375 minutes]:
sync.runtime_notifyListWait(0xc0004089d0, 0x54)
	runtime/sema.go:513 +0x13d
sync.(*Cond).Wait(0x47474f4c01656d61?)
	sync/cond.go:56 +0x8c
github.com/VictoriaMetrics/VictoriaMetrics/lib/pacelimiter.(*PaceLimiter).WaitIfNeeded(0xc00030e660)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/pacelimiter/pacelimiter.go:52 +0x65
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.checkSearchDeadlineAndPace(...)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/search.go:466
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexSearch).updateMetricIDsForOrSuffix(0xc106302000, {0xc190a773b0, 0x46, 0x50}, 0xc0e42944e0, 0xe4e1f2, 0x7fffffffffffffff)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/index_db.go:2433 +0x267
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexSearch).updateMetricIDsForOrSuffixes(0xc10a137218?, 0xc286af2370, 0xc0e42944e0, 0xe4e1f2, 0x7fffffffffffffff)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/index_db.go:2412 +0x3b9
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexSearch).getMetricIDsForTagFilter(0xbfc2a0?, 0xc286af2370, 0x890640?, 0x1?)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/index_db.go:2293 +0xab
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexSearch).getMetricIDsForDateTagFilter(0xc106302000, 0xc11077a200, 0x4ab5, {0xc248233450, 0x9, 0x10}, 0x6b2caf?, 0x7fffffffffffffff)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/index_db.go:2853 +0x753
github.com/VictoriaMetrics/VictoriaMetrics/lib/storage.(*indexSearch).getMetricIDsForDateAndFilters(0xc106302000, 0x4ab5, 0xc0d5e7a1c0, 0x493e1)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/storage/index_db.go:2583 +0xb5b

I also snapshot of  the pprof, most of goroutine are blocked by select channel:
![image](https://user-images.githubusercontent.com/103412360/168347759-f0bab6e5-b38e-4e92-a758-b92c49f3ffb1.png)

**Comment by valyala** [comment]
> the concurrentDialsCh only limits the rate to handshakeFunc, is there anyway to limit the max num of connection?

There is no strict limit on the number of established connections from vmselect to vmstorage mostly because it is expected that the number of connections is proportional to the number of concurrent queries (which is limited by 16 as explained [above](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2552#issuecomment-1123654677) ). It is also expected that the number of connections per each query rarely exceeds 10. It looks like this assumption doesn't work in your case according to:

> we have a very slow query which contains 100 subqueries join by 'or' statement.

and

> I find there is a query with more than 10k subquery... this query must be the problem

@eddietanshuo , it would be great if you could provide a sample of these queries, since I cannot reproduce high number of connections ( `vm_tcpdialer_conns` and `vm_tcpdialer_conns_idle` metrics ) with queries like `foo or bar or ... or abc`. VictoriaMetrics performs sequential search for the listed time series selectors in the query, so the query uses only a single connection per each known vmstorage node.


> By the way, I find some goroutine are blocked by pacelimiter with conditional variable. They seems never wakeup

Could you share a screenshot for the following queries:

1. `rate(vm_search_delays_total)`
2. `rate(process_cpu_seconds_total) / process_cpu_cores_available`

Probably there is too high data ingestion rate for the given CPU resources at `vmstorage`. This can lead to a livelock at https://github.com/VictoriaMetrics/VictoriaMetrics/blob/6bc52fe41a7cf389606cb8e742f27d7b08f7a6c1/lib/pacelimiter/pacelimiter.go#L50-L53 This may prevent from query execution, because VictoriaMetrics prioritizes data ingestion over query execution when there is no enough CPU resources.

**Comment by valyala** [comment]
FYI, the commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/d8a276fbe471527b8da43a0fb7cb92321b2d982e is included in [VictoriaMetrics cluster v1.77.2](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).

**Comment by eddietanshuo** [comment]
Thanks for reply,

1. For the livelock problem, cpu is very high at that time, since the node has 200k connections. We only reserve 3 days data, so I can not find that screenshot. I  have a screenshot for addrow_capacity, it seems that the node's ingesting has been stale.
![image](https://user-images.githubusercontent.com/103412360/169631199-7cafdfeb-1edc-434e-8d9d-17d4f01c64db.png)
After we remove the storage node from vmselect, the node's load reduce and start to process ingesting, and all livelock has been released. I thinks this case is also related to high connection rate.

2. For the query problem, I only can provide more context:
We use an external cmdb to store all servers' ip addr and it's department info.  For example someone in a big department lookup the cpu usage of the whole department, our service first search the cmdb and obtain 10000 server's ip. Then we use metricsql to join all these instances: `cpu_usage{addr="ipA"} or cpu_usage{addr="ipB"} ` to get the resource usage from timeseries db. And we also limited the string length of this query, we slice the query to make length less than 10k. 

We optimized the query to be `cpu_usage{addr="ipA|ipB"}` and the problem solved. Great thanks for help.

**Comment by eddietanshuo** [comment]
sorry, to make a correct:
The query is `label_replace(irate(cpu_usage{addr="ipA"}[60s]), "__name__":"cpu_usage") or label_replace(irate(cpu_usage{addr="ipB"}[60s]), "__name__":"cpu_usage")` not  `cpu_usage{addr="ipA"} or cpu_usage{addr="ipB"}`

**Comment by valyala** [comment]
> After we remove the storage node from vmselect, the node's load reduce and start to process ingesting, and all livelock has been released. I thinks this case is also related to high connection rate.

@eddietanshuo , try upgrading VictoriaMetrics cluster components to [v1.78.1](https://docs.victoriametrics.com/CHANGELOG.html#v1781) - it contains an enhancement for detecting unreachable `vmstorage` nodes, which may improve query performance in this case - see [this comment](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/711#issuecomment-1160363187).

**Comment by eddietanshuo** [comment]
hello @valyala 



After upgrade to [v1.78.1](https://docs.victoriametrics.com/CHANGELOG.html#v1781), we meet this issue again. When an assisted merge happens, the vmselect connections for the problem vmstorage will be held by the pacelimiter, and the tcp connections for the connection will reach 30k. At this moment the assisted merge will also slow down during the high load avg (we try to connect the node by ssh, it also take us more than 1 minute) on this node. We also find the conection on this node closed as soon as it established.

It's possible to limit the connection pool size of vmselect (since we don't need many long rpc connections)?  Or just give a max value for the pace limiter and fast return error when limit reached? 



![image](https://user-images.githubusercontent.com/103412360/182540904-b73cb154-4ea9-4d53-a779-38b8512d297e.png)


![image](https://user-images.githubusercontent.com/103412360/182540656-2c16b189-2c8f-4b29-a004-ef4312f08990.png)


![image](https://user-images.githubusercontent.com/103412360/182540424-bcfe9a9b-f7f9-4660-a0c8-f6114b243c8c.png)

**Comment by eddietanshuo** [comment]
@valyala it looks like a system design problem because the system could not recover until we restart all vmselect. 
![image](https://user-images.githubusercontent.com/103412360/182543586-26cba682-2748-4c1a-8052-51d489e5b005.png)

**Comment by hagen1778** [comment]
Hello @eddietanshuo!
Looks like the change from https://github.com/VictoriaMetrics/VictoriaMetrics/issues/711#issuecomment-1160363187 supposed to prevent vmselect from establishing too many connections and mark bad vmstorage as "dead" much faster.
Have you tried the change by updating to [1.78](https://docs.victoriametrics.com/CHANGELOG.html#v1780) or higher?

**Comment by eddietanshuo** [comment]
Hello @hagen1778 
Our version is 1.79.3-cluster now.  During some condition it's still locked and result in too many vmselect connection, for  example:  one of node's disk ioutil is full, then the node's merge speed is down, and result in vmstorage's ingest rate becoming slow, as an effect of pacelimiter, the connection from vmselect is locked and accumulated in backend. In this case the cpu usage of that node is very low like 50% of normal, but the gc time reach 500ms, since a lot of decompress works running in the backend. One of my suggestion is to limit vmselect rpc connection.

**Comment by eddietanshuo** [comment]
By the way we figure out the root cause:
We used lvm to manager the disk. We have 8 hdd disk device and use lvm to pack them to a large volume and create a fs on it. However the lvm is not very suitable to do this since we find in some condition the parts of one merge process may storage in the same disk device and the output is also in this disk, so the ioutil is soon become 80%+ and the merge speed become very slow. This problem has been solved by using raid.
